apis/apiextensions/v1/composition_types.go

```diff
@@ -27,38 +27,59 @@ type CompositionSpec struct {
 	// +immutable
 	CompositeTypeRef TypeReference `json:"compositeTypeRef"`
 
-	// PatchSets define a named set of patches that may be included by
-	// any resource in this Composition.
-	// PatchSets cannot themselves refer to other PatchSets.
+	// Mode controls what type or "mode" of Composition will be used.
+	//
+	// "Resources" (the default) indicates that a Composition uses what is
+	// commonly referred to as "Patch & Transform" or P&T composition. This mode
+	// of Composition uses an array of resources, each a template for a composed
+	// resource.
+	//
+	// "Pipeline" indicates that a Composition specifies a pipeline
+	// of Composition Functions, each of which is responsible for producing
+	// composed resources that Crossplane should create or update. THE PIPELINE
+	// MODE IS A BETA FEATURE. It is not honored if the relevant Crossplane
+	// feature flag is disabled.
+	// +optional
+	// +kubebuilder:validation:Enum=Resources;Pipeline
+	// +kubebuilder:default=Resources
+	Mode *CompositionMode `json:"mode,omitempty"`
```

Comment:
I assume this field would be immutable, i.e. it won't be possible to switch between composition modes?

If yes, should we mark it as immutable with CEL: https://kubernetes.io/blog/2022/09/29/enforce-immutability-using-cel/

Reply:
Wouldn't making it immutable get the migration harder from one mode to the other?

Reply:
I think we do want to support switching modes. We could make `mode` immutable within a Composition, but it's still possible to change what Composition your XR uses. So you could switch from a `mode: Resources` Composition to a `mode: Pipeline` one.

I do think we should make it possible to migrate existing XRs to use Functions instead of P&T. Requiring folks to (re)create new XRs seems like it would hurt adoption.